internal/mesh/internal/controllers/sidecarproxy/cache/cache.go

```diff
@@ -0,0 +1,144 @@
+package cache
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/consul/internal/mesh/internal/types/intermediate"
+	"github.com/hashicorp/consul/internal/resource"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+// Cache stores information needed for the sidecar-proxy controller to reconcile efficiently.
+// This currently means storing a list of all destinations for easy look up
+// as well as indices of source proxies where those destinations are referenced.
+//
+// It is the responsibility of the controller and its subcomponents (like mapper and data fetcher)
+// to keep this cache up-to-date as we're observing new data.
+type Cache struct {
+	lock sync.RWMutex
+
+	// store is a map from destination service reference and port as a string ID
+	// to the object representing destination reference.
+	store map[string]*intermediate.CombinedDestinationRef
+
+	// sourceProxiesIndex stores a map from a string representation of source proxy ID
+	// to the keys in the store map.
+	sourceProxiesIndex map[string]storeKeys
+}
+
+type storeKeys map[string]struct{}
+
+func New() *Cache {
+	return &Cache{
+		store:              make(map[string]*intermediate.CombinedDestinationRef),
+		sourceProxiesIndex: make(map[string]storeKeys),
+	}
+}
+
+func KeyFromID(id *pbresource.ID) string {
+	return fmt.Sprintf("%s/%s/%s",
+		resource.ToGVK(id.Type),
+		resource.TenancyToString(id.Tenancy),
+		id.Name)
+}
+
+func KeyFromRefAndPort(ref *pbresource.Reference, port string) string {
+	return fmt.Sprintf("%s:%s",
+		resource.ReferenceToString(ref),
+		port)
+}
+
+func (c *Cache) Write(d *intermediate.CombinedDestinationRef) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	key := KeyFromRefAndPort(d.ServiceRef, d.Port)
+
+	c.store[key] = d
```

Comment:
I suspect you have the same problem that I encountered in bimapper. You need to clear the old source proxies references before populating new ones. [reference](https://github.com/hashicorp/consul/blob/main/internal/resource/mappers/bimapper/bimapper.go#L102)

For you that would mean turning the body of this into something like `writeLocked`, converting the body of `Delete` into `deleteLocked` and then having `Write` acquire the lock and call `deleteLocked` followed by `writeLocked` to properly clear the old references before replacing them with new ones.